File: capy/src/e2e-coo/py/dev/capylang/test/object_construction_interop_test.py
import contextlib
import io
import os
import pathlib
import sys
import unittest


generated_dir = pathlib.Path(os.environ["CAPY_E2E_COO_PY_GENERATED_DIR"])
sys.path.insert(0, str(generated_dir))

import dev.capylang.test.ObjectConstructionInterop as interop
from dev.capylang.test.EmptyTracked import EmptyTracked
from dev.capylang.test.TrackedPerson import TrackedPerson


class ObjectConstructionInteropPythonE2ETest(unittest.TestCase):
    def capture(self, callback):
        out = io.StringIO()
        with contextlib.redirect_stdout(out):
            result = callback()
        return result, out.getvalue()

    def test_functional_object_construction_is_lazy_and_fresh(self):
        effect, output = self.capture(lambda: interop.makePerson("Ada"))
        self.assertEqual(output, "")

        first, output = self.capture(effect.unsafe_run)
        self.assertIsInstance(first, TrackedPerson)
        self.assertEqual(first.label(), "Ada")
        self.assertEqual(output, "constructed:Ada\n")

        second, output = self.capture(effect.unsafe_run)
        self.assertIsInstance(second, TrackedPerson)
        self.assertIsNot(second, first)
        self.assertEqual(output, "constructed:Ada\n")

    def test_parent_zero_arg_and_sequence_cases(self):
        printable, _ = self.capture(lambda: interop.makePrintable("Bara").unsafe_run())
        self.assertIsInstance(printable, TrackedPerson)
        self.assertEqual(printable.label(), "Bara")

        empty_effect, output = self.capture(interop.makeEmpty)
        self.assertEqual(output, "")
        empty, output = self.capture(empty_effect.unsafe_run)
        self.assertIsInstance(empty, EmptyTracked)
        self.assertEqual(empty.label(), "empty")
        self.assertEqual(output, "empty\n")

        result, output = self.capture(lambda: interop.sequenceTwo("A", "B").unsafe_run())
        self.assertEqual(result, "A:B")
        self.assertEqual(output, "constructed:A\nconstructed:B\n")

File: capy/src/e2e-coo/js/dev/capylang/test/ObjectConstructionInterop.test.js
const test = require('node:test');
const { assert, generatedModule, captureOutput } = require('./_support');

test('functional object construction is lazy, fresh, and CommonJS-wired', () => {
    const interop = generatedModule('dev/capylang/test/ObjectConstructionInterop.js');
    const { TrackedPerson } = generatedModule('dev/capylang/test/TrackedPerson.js');

    const created = captureOutput(() => interop.makePerson('Ada'));
    assert.equal(created.stdout, '');

    const first = captureOutput(() => created.result.unsafe_run());
    assert.equal(first.result instanceof TrackedPerson, true);
    assert.equal(first.result.label(), 'Ada');
    assert.equal(first.stdout, 'constructed:Ada\n');

    const second = captureOutput(() => created.result.unsafe_run());
    assert.equal(second.result instanceof TrackedPerson, true);
    assert.notEqual(second.result, first.result);
    assert.equal(second.stdout, 'constructed:Ada\n');
});

test('functional object construction supports parent types, zero args, and sequencing', () => {
    const interop = generatedModule('dev/capylang/test/ObjectConstructionInterop.js');
    const { TrackedPerson } = generatedModule('dev/capylang/test/TrackedPerson.js');
    const { EmptyTracked } = generatedModule('dev/capylang/test/EmptyTracked.js');

    const printable = captureOutput(() => interop.makePrintable('Bara').unsafe_run());
    assert.equal(printable.result instanceof TrackedPerson, true);
    assert.equal(printable.result.label(), 'Bara');

    const emptyEffect = captureOutput(() => interop.makeEmpty());
    assert.equal(emptyEffect.stdout, '');
    const empty = captureOutput(() => emptyEffect.result.unsafe_run());
    assert.equal(empty.result instanceof EmptyTracked, true);
    assert.equal(empty.result.label(), 'empty');
    assert.equal(empty.stdout, 'empty\n');

    const sequenced = captureOutput(() => interop.sequenceTwo('A', 'B').unsafe_run());
    assert.equal(sequenced.result, 'A:B');
    assert.equal(sequenced.stdout, 'constructed:A\nconstructed:B\n');
});
